src/threadpool.c

```diff
@@ -176,11 +202,43 @@ void uv__work_submit(uv_loop_t* loop,
                      struct uv__work* w,
                      void (*work)(struct uv__work* w),
                      void (*done)(struct uv__work* w, int status)) {
+  int err;
+
   uv_once(&once, init_once);
   w->loop = loop;
   w->work = work;
   w->done = done;
-  post(&w->wq);
+
+  uv_mutex_lock(&mutex);
+
+  if (busy_threads >= num_threads && num_threads < max_threads) {
+    /* Drop the lock for a moment, starting a thread can be slow. */
+    uv_mutex_unlock(&mutex);
+    err = new_detached_thread();
+    uv_mutex_lock(&mutex);
+
+    /* There is technically a race window between new_detached_thread() where
+     * the new thread exits and decrements num_threads before we increment it
+     * here.  It's a benign race, though; worst case, num_threads is less than
+     * zero for a short time.  We just take that in stride, the important part
+     * is that num_threads > 0 means there is a thread to service the work item.
+     */
+    if (err == 0)
+      num_threads += 1;
+  }
+
+  if (num_threads > 0) {
+    QUEUE_INSERT_TAIL(&wq, &w->wq);
+    uv_cond_signal(&cond);
+    uv_mutex_unlock(&mutex);
+  } else {
+    /* If we can't start a thread, fall back to doing the work synchronously.
```

Comment:
Returning an error means all call sites need to be updated to handle errors.

The hairy part is cleanup.  There will be memory leaks all over the place unless fs.c, getaddrinfo.c and getnameinfo.c go through extensive rewrites.